applications/StructuralMechanicsApplication/custom_processes/advance_in_time_strategy_high_cycle_fatigue_process.cpp

```diff
@@ -0,0 +1,251 @@
+/***********************************************************************************/
+
+void AdvanceInTimeStrategyHighCycleFatigueProcess::Execute()
+{
+    
+    auto& process_info = mrModelPart.GetProcessInfo();
+    bool cycle_found = false;
+    std::vector<bool> cycle_identificator;
+    std::vector<double> damage;
+    process_info[ADVANCE_STRATEGY_APPLIED] = false;
+
+    if (process_info[DAMAGE_ACTIVATION] == false) {
+        for (auto& r_elem : mrModelPart.Elements()) {
+            r_elem.CalculateOnIntegrationPoints(DAMAGE, damage, process_info);
+            const unsigned int number_of_ip = r_elem.GetGeometry().IntegrationPoints(r_elem.GetIntegrationMethod()).size();
+            for (unsigned int i = 0; i < number_of_ip; i++){
+                    if (damage[i] > 0.0){
+                        process_info[DAMAGE_ACTIVATION] = true;
+                    }
+            }
+        }
+    }
+
+    this->CyclePeriodPerIntegrationPoint(cycle_found);  //This method detects if a cycle has finished somewhere in the model and 
+                                                        //computes the time period of the cycle that has just finished.
+    
+    if (cycle_found) {  //If a cycle has finished then it is possible to apply the advancing strategy
+        bool advancing_strategy = false;
+        this->StableConditionForAdvancingStrategy(advancing_strategy, process_info[DAMAGE_ACTIVATION]);  //Check if the conditions are optimal to apply the advancing strategy in 
+                                                                        //terms of max stress and reversion factor variation.  
+        double increment = 0.0;
+        if (advancing_strategy) {
+            if (!process_info[DAMAGE_ACTIVATION]) {
+                this->TimeIncrement(increment);
+                if(increment > 0){
+                    this->TimeAndCyclesUpdate(increment);
+                    process_info[ADVANCE_STRATEGY_APPLIED] = true;
+                }
+            } else {
+                double increment = mThisParameters["fatigue"]["advancing_strategy_damage"].GetDouble();
+                this->TimeAndCyclesUpdate(increment);   
+                process_info[ADVANCE_STRATEGY_APPLIED] = true;             
+            }
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void AdvanceInTimeStrategyHighCycleFatigueProcess::CyclePeriodPerIntegrationPoint(bool& rCycleFound) 
+{
+    auto& process_info = mrModelPart.GetProcessInfo();
+    std::vector<bool> cycle_identificator;
+    std::vector<double> previous_cycle_time;    //time when the previous cycle finished. It is used to obtain the new period for the current cycle
+    std::vector<double> period;
+    double time = process_info[TIME];
+    
+    for (auto& r_elem : mrModelPart.Elements()) {
```

Comment:
please use the Parallel Utilities to make this loop parallel

Reply:
I am not familiarized with the OpenMP tools, I would like to address it on the future.